include/wx/webrequest.h

```diff
@@ -0,0 +1,305 @@
+///////////////////////////////////////////////////////////////////////////////
+// Name:        wx/webrequest.h
+// Purpose:     wxWebRequest base classes
+// Author:      Tobias Taschner
+// Created:     2018-10-17
+// Copyright:   (c) 2018 wxWidgets development team
+// Licence:     wxWindows licence
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef _WX_WEBREQUEST_H
+#define _WX_WEBREQUEST_H
+
+#include "wx/defs.h"
+
+#if wxUSE_WEBREQUEST
+
+#include "wx/event.h"
+#include "wx/ffile.h"
+#include "wx/object.h"
+#include "wx/scopedptr.h"
+#include "wx/sharedptr.h"
+#include "wx/stream.h"
+#include "wx/vector.h"
+#include "wx/versioninfo.h"
+
+class wxWebResponse;
+class wxWebSession;
+class wxWebAuthChallenge;
+
+WX_DECLARE_STRING_HASH_MAP(wxString, wxWebRequestHeaderMap);
+
+class WXDLLIMPEXP_NET wxWebRequest : public wxEvtHandler, public wxRefCounter
+{
+public:
+    enum State
+    {
+        State_Idle,
+        State_Unauthorized,
+        State_Active,
+        State_Completed,
+        State_Failed,
+        State_Cancelled
+    };
+
+    enum Storage
+    {
+        Storage_Memory,
+        Storage_File,
+        Storage_None
+    };
+
+    virtual ~wxWebRequest() { }
+
+    virtual void SetHeader(const wxString& name, const wxString& value)
+    { m_headers[name] = value; }
+
+    virtual void SetMethod(const wxString& method) { m_method = method; }
+
+    void SetData(const wxString& text, const wxString& contentType, const wxMBConv& conv = wxConvUTF8);
+
+    void SetData(wxSharedPtr<wxInputStream> dataStream, const wxString& contentType, wxFileOffset dataSize = wxInvalidOffset);
+
+    void SetIgnoreServerErrorStatus(bool ignore) { m_ignoreServerErrorStatus = ignore; }
```

Comment:
OK, but would it actually hurt to download the body even in the case (1)? I don't think so, especially because it will usually be empty (or small) anyhow. Which is why I think this method is not really needed, you can just ignore the body if you don't need it, but IMO we should always receive the entire server response in any case.